executor/read/nested_pattern_executor.rs

```diff
@@ -0,0 +1,267 @@
+        match self {
+            Self::Identity(mapper) => mapper.prepare_and_map_input(input),
+            Self::Negation(mapper) => mapper.prepare_and_map_input(input),
+            Self::InlinedFunction(mapper) => mapper.prepare_and_map_input(input),
+            Self::Offset(mapper) => mapper.prepare_and_map_input(input),
+            Self::Limit(mapper) => mapper.prepare_and_map_input(input),
+        }
+    }
+
+    pub(super) fn map_output(
+        &mut self,
+        input: &MaybeOwnedRow<'static>,
+        subquery_result: Option<FixedBatch>,
+    ) -> SubQueryResult {
+        match self {
+            Self::Identity(mapper) => mapper.map_output(input, subquery_result),
+            Self::Negation(mapper) => mapper.map_output(input, subquery_result),
+            Self::InlinedFunction(mapper) => mapper.map_output(input, subquery_result),
+            Self::Offset(mapper) => mapper.map_output(input, subquery_result),
+            Self::Limit(mapper) => mapper.map_output(input, subquery_result),
+        }
+    }
+}
+
+impl NestedPatternExecutor {
+    pub(super) fn to_parts_mut(&mut self) -> (&mut [NestedPatternBranch], SubQueryResultMapperEnumMut<'_>) {
+        match self {
+            NestedPatternExecutor::Disjunction(branches, mapper) => {
+                (branches.as_mut_slice(), SubQueryResultMapperEnumMut::Identity(mapper))
+            }
+            NestedPatternExecutor::Negation(branch, mapper) => {
+                (branch.as_mut_slice(), SubQueryResultMapperEnumMut::Negation(mapper))
+            }
+            NestedPatternExecutor::InlinedFunction(branch, mapper) => {
+                (branch.as_mut_slice(), SubQueryResultMapperEnumMut::InlinedFunction(mapper))
+            }
+            NestedPatternExecutor::Offset(branch, mapper) => {
+                (branch.as_mut_slice(), SubQueryResultMapperEnumMut::Offset(mapper))
+            }
+            NestedPatternExecutor::Limit(branch, mapper) => {
+                (branch.as_mut_slice(), SubQueryResultMapperEnumMut::Limit(mapper))
+            }
+        }
+    }
+}
+
+pub(super) struct NestedPatternBranch {
+    pub(super) input: Option<MaybeOwnedRow<'static>>,
+    pub(super) pattern_executor: PatternExecutor,
+}
+
+impl NestedPatternBranch {
+    pub(crate) fn new(pattern_executor: PatternExecutor) -> Self {
+        Self { input: None, pattern_executor }
+    }
+
+    pub(super) fn prepare(&mut self, input: MaybeOwnedRow<'static>) {
+        self.input = Some(input.clone());
+        self.pattern_executor.prepare(FixedBatch::from(input.clone()));
+    }
+}
+
+pub(super) trait SubQueryResultMapper {
+    fn prepare_and_map_input(&mut self, input: &MaybeOwnedRow<'_>) -> MaybeOwnedRow<'static>;
+    fn map_output(&mut self, input: &MaybeOwnedRow<'_>, subquery_result: Option<FixedBatch>) -> SubQueryResult;
+}
+
+pub(super) enum SubQueryResult {
+    Retry(Option<FixedBatch>),
+    Done(Option<FixedBatch>),
+}
+
+pub(super) struct NegationMapper;
```

Comment:
It feels like NegationMapper shouldn't use the default `map_output`, which clones the input unnecessarily